File: sdk/python/openclaw_relay/channel.py
"""Layer 0: WebSocket framing and channel management."""

from __future__ import annotations

import asyncio
import json
import logging
import random
from typing import Any

import websockets
import websockets.exceptions

logger = logging.getLogger(__name__)

# Reconnection parameters
_INITIAL_BACKOFF = 1.0  # seconds
_MAX_BACKOFF = 60.0  # seconds
_JITTER_FACTOR = 0.25


class ChannelError(Exception):
    """Raised when the relay returns an error frame."""


class ChannelReconnected(Exception):
    """Raised by recv() after the channel reconnects at Layer 0.

    Upper layers should treat this as a signal to rebuild their session
    (e.g. redo HELLO handshake and derive fresh encryption keys).
    """


class ChannelConnection:
    """Layer 0: WebSocket connection to an OpenClaw relay server.

    Handles connect/register/join lifecycle, frame serialisation and
    automatic reconnection with exponential back-off + jitter.
    """

    def __init__(self) -> None:
        self._ws: websockets.WebSocketClientProtocol | None = None
        self._relay_url: str | None = None
        self._closed: bool = False

        # Reconnection state
        self._backoff: float = _INITIAL_BACKOFF
        self._reconnect_lock = asyncio.Lock()

        # Populated after register/join so that reconnection can re-establish
        self._role: str | None = None  # "gateway" or "client"
        self._channel_hash: str | None = None
        self._client_id: str | None = None

    # ------------------------------------------------------------------
    # Connection
    # ------------------------------------------------------------------

    async def connect(self, relay_url: str) -> None:
        """Open a WebSocket connection to *relay_url*."""
        self._relay_url = relay_url
        self._closed = False
        await self._do_connect()

    async def _do_connect(self) -> None:
        """Internal: establish the raw WebSocket."""
        assert self._relay_url is not None
        self._ws = await websockets.connect(
            self._relay_url,
            ping_interval=30,
            ping_timeout=10,
            close_timeout=5,
        )
        self._backoff = _INITIAL_BACKOFF
        logger.info("WebSocket connected to %s", self._relay_url)

    # ------------------------------------------------------------------
    # Channel lifecycle
    # ------------------------------------------------------------------

    async def register(self, channel_hash: str) -> dict:
        """Register as a *gateway* for *channel_hash*.

        Returns the ``registered`` acknowledgement frame.
        """
        self._role = "gateway"
        self._channel_hash = channel_hash

        await self._send_frame({"type": "register", "channel": channel_hash, "version": 1})
        frame = await self._recv_frame_expect("registered")
        return frame

    async def join(self, channel_hash: str, client_id: str) -> dict:
        """Join a channel as a *client*.

        Returns the ``joined`` acknowledgement frame.
        """
        self._role = "client"
        self._channel_hash = channel_hash
        self._client_id = client_id

        await self._send_frame({
            "type": "join",
            "channel": channel_hash,
            "version": 1,
            "client_id": client_id,
        })
        frame = await self._recv_frame_expect("joined")
        return frame

    # ------------------------------------------------------------------
    # Data exchange
    # ------------------------------------------------------------------

    async def send_data(self, to: str, payload: str) -> None:
        """Send a ``data`` frame to peer *to*."""
        await self._send_frame({
            "type": "data",
            "to": to,
            "payload": payload,
        })

    async def recv(self) -> dict:
        """Receive and return the next parsed frame.

        Automatically handles ``ping`` frames from the server by replying
        with ``pong``, and transparently reconnects on connection loss.
        """
        while True:
            try:
                frame = await self._recv_raw()
            except (
                websockets.exceptions.ConnectionClosed,
                websockets.exceptions.ConnectionClosedError,
                websockets.exceptions.ConnectionClosedOK,
                ConnectionError,
                OSError,
            ):
                if self._closed:
                    raise ConnectionError("Connection closed")
                await self._reconnect()
                raise ChannelReconnected()

            frame_type = frame.get("type")

            # Respond to server pings transparently
            if frame_type == "ping":
                await self._send_frame({"type": "pong"})
                continue

            if frame_type == "error":
                logger.error("Relay error: %s", frame.get("message", frame))

            return frame

    async def send_ping(self) -> None:
        """Send a client-initiated ``ping`` frame."""
        await self._send_frame({"type": "ping"})

    async def close(self) -> None:
        """Cleanly close the WebSocket."""
        self._closed = True
        if self._ws is not None:
            try:
                await self._ws.close()
            except Exception:
                pass
            self._ws = None
        logger.info("Channel connection closed")

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    async def _send_frame(self, frame: dict[str, Any]) -> None:
        """Serialise *frame* as JSON and send over the WebSocket."""
        if self._ws is None:
            raise ConnectionError("Not connected")
        raw = json.dumps(frame, separators=(",", ":"))
        await self._ws.send(raw)

    async def _recv_raw(self) -> dict:
        """Receive a single raw frame (JSON) from the WebSocket."""
        if self._ws is None:
            raise ConnectionError("Not connected")
        raw = await self._ws.recv()
        if isinstance(raw, bytes):
            raw = raw.decode()
        return json.loads(raw)

    async def _recv_frame_expect(self, expected_type: str) -> dict:
        """Receive frames until one matches *expected_type*.

        If an ``error`` frame arrives instead, raise ``ChannelError``.
        Server ``ping`` frames are answered transparently.
        """
        while True:
            frame = await self._recv_raw()
            ft = frame.get("type")
            if ft == expected_type:
                return frame
            if ft == "ping":
                await self._send_frame({"type": "pong"})
                continue
            if ft == "error":
                raise ChannelError(frame.get("message", str(frame)))
            # Ignore unexpected frame types during the handshake
            logger.debug("Ignoring unexpected frame during %s wait: %s", expected_type, ft)

    # ------------------------------------------------------------------
    # Reconnection with exponential back-off + jitter
    # ------------------------------------------------------------------

    async def _reconnect(self) -> None:
        """Attempt to reconnect with exponential back-off."""
        async with self._reconnect_lock:
            # Another coroutine may have reconnected while we waited for the lock.
            if self._ws is not None:
                try:
                    await self._ws.ping()
                    return
                except Exception:
                    pass

            while not self._closed:
                jitter = random.uniform(0, _JITTER_FACTOR * self._backoff)
                delay = self._backoff + jitter
                logger.info("Reconnecting in %.1fs ...", delay)
                await asyncio.sleep(delay)

                try:
                    await self._do_connect()
                except Exception as exc:
                    logger.warning("Reconnect failed: %s", exc)
                    self._backoff = min(self._backoff * 2, _MAX_BACKOFF)
                    continue

                # Re-establish channel membership
                try:
                    if self._role == "gateway" and self._channel_hash:
                        await self.register(self._channel_hash)
                    elif self._role == "client" and self._channel_hash and self._client_id:
                        await self.join(self._channel_hash, self._client_id)
                except Exception as exc:
                    logger.warning("Re-join after reconnect failed: %s", exc)
                    self._backoff = min(self._backoff * 2, _MAX_BACKOFF)
                    continue

                logger.info("Reconnected successfully")
                return

            raise ConnectionError("Connection closed, cannot reconnect")

File: sdk/js/src/channel.ts
import WebSocket from 'ws';

const INITIAL_BACKOFF_MS = 1000;
const MAX_BACKOFF_MS = 60000;
const JITTER_FACTOR = 0.25;

class ChannelDisconnected extends Error {
  constructor(message = 'channel disconnected') {
    super(message);
    this.name = 'ChannelDisconnected';
  }
}

class AsyncQueue<T> {
  private items: T[] = [];
  private waiters: Array<{ resolve: (value: T) => void; reject: (err: Error) => void }> = [];
  private failure: Error | null = null;

  push(item: T): void {
    if (this.failure) return;
    const waiter = this.waiters.shift();
    if (waiter) {
      waiter.resolve(item);
      return;
    }
    this.items.push(item);
  }

  fail(error: Error): void {
    if (this.failure) return;
    this.failure = error;
    while (this.waiters.length > 0) {
      const waiter = this.waiters.shift();
      if (waiter) waiter.reject(error);
    }
  }

  async shift(): Promise<T> {
    if (this.failure) throw this.failure;
    if (this.items.length > 0) {
      return this.items.shift() as T;
    }
    return await new Promise<T>((resolve, reject) => {
      this.waiters.push({ resolve, reject });
    });
  }
}

export class ChannelError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ChannelError';
  }
}

export class ChannelReconnected extends Error {
  constructor() {
    super('channel reconnected');
    this.name = 'ChannelReconnected';
  }
}

export class ChannelConnection {
  private ws: WebSocket | null = null;
  private relayUrl: string | null = null;
  private closed = false;
  private backoffMs = INITIAL_BACKOFF_MS;
  private reconnectPromise: Promise<void> | null = null;
  private role: 'gateway' | 'client' | null = null;
  private channelHash: string | null = null;
  private clientId: string | null = null;
  private incoming = new AsyncQueue<Record<string, unknown>>();

  async connect(relayUrl: string): Promise<void> {
    this.relayUrl = relayUrl;
    this.closed = false;
    await this._doConnect();
  }

  async register(channelHash: string): Promise<Record<string, unknown>> {
    this.role = 'gateway';
    this.channelHash = channelHash;
    await this._sendFrame({ type: 'register', channel: channelHash, version: 1 });
    return await this._recvFrameExpect('registered');
  }

  async join(channelHash: string, clientId: string): Promise<Record<string, unknown>> {
    this.role = 'client';
    this.channelHash = channelHash;
    this.clientId = clientId;
    await this._sendFrame({ type: 'join', channel: channelHash, version: 1, client_id: clientId });
    return await this._recvFrameExpect('joined');
  }

  async sendData(to: string, payload: string): Promise<void> {
    await this._sendFrame({ type: 'data', to, payload });
  }

  async recv(): Promise<Record<string, unknown>> {
    while (true) {
      let frame: Record<string, unknown>;
      try {
        frame = await this.incoming.shift();
      } catch (err) {
        if (err instanceof ChannelDisconnected) {
          if (this.closed) {
            throw new Error('Connection closed');
          }
          await this._reconnect();
          throw new ChannelReconnected();
        }
        throw err;
      }

      const frameType = String(frame.type || '');
      if (frameType === 'ping') {
        await this._sendFrame({ type: 'pong' });
        continue;
      }

      return frame;
    }
  }

  async sendPing(): Promise<void> {
    await this._sendFrame({ type: 'ping' });
  }

  async close(): Promise<void> {
    this.closed = true;
    this.incoming.fail(new ChannelDisconnected('closed'));
    if (this.ws) {
      try {
        this.ws.close();
      } catch {
        // ignore
      }
      this.ws = null;
    }
  }

  private async _sendFrame(frame: Record<string, unknown>): Promise<void> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      throw new Error('WebSocket not connected');
    }
    this.ws.send(JSON.stringify(frame));
  }

  private async _recvFrameExpect(expected: string): Promise<Record<string, unknown>> {
    while (true) {
      const frame = await this.incoming.shift();
      const frameType = String(frame.type || '');
      if (frameType === expected) return frame;
      if (frameType === 'ping') {
        await this._sendFrame({ type: 'pong' });
        continue;
      }
      if (frameType === 'error') {
        const code = String(frame.code || 'error');
        const message = String(frame.message || frameType);
        throw new ChannelError(`[${code}] ${message}`);
      }
    }
  }

  private async _doConnect(): Promise<void> {
    if (!this.relayUrl) {
      throw new Error('Relay URL not set');
    }

    const ws = new WebSocket(this.relayUrl, { handshakeTimeout: 10000 });
    this.ws = ws;
    this.incoming = new AsyncQueue<Record<string, unknown>>();

    ws.on('message', (data: WebSocket.RawData) => {
      try {
        const raw = typeof data === 'string' ? data : data.toString();
        const frame = JSON.parse(raw) as Record<string, unknown>;
        this.incoming.push(frame);
      } catch {
        // Ignore malformed frames
      }
    });

    ws.on('close', () => {
      if (!this.closed) {
        this.incoming.fail(new ChannelDisconnected('closed'));
      }
    });

    ws.on('error', () => {
      if (!this.closed) {
        this.incoming.fail(new ChannelDisconnected('error'));
      }
    });

    await new Promise<void>((resolve, reject) => {
      ws.once('open', () => resolve());
    ws.once('error', (err: Error) => reject(err));
    });

    this.backoffMs = INITIAL_BACKOFF_MS;
  }

  private async _reconnect(): Promise<void> {
    if (this.reconnectPromise) {
      return await this.reconnectPromise;
    }

    this.reconnectPromise = (async () => {
      while (!this.closed) {
        const jitter = Math.random() * JITTER_FACTOR * this.backoffMs;
        const delay = this.backoffMs + jitter;
        await new Promise((resolve) => setTimeout(resolve, delay));

        try {
          await this._doConnect();
        } catch {
          this.backoffMs = Math.min(this.backoffMs * 2, MAX_BACKOFF_MS);
          continue;
        }

        try {
          if (this.role === 'gateway' && this.channelHash) {
            await this.register(this.channelHash);
          } else if (this.role === 'client' && this.channelHash && this.clientId) {
            await this.join(this.channelHash, this.clientId);
          }
        } catch {
          this.backoffMs = Math.min(this.backoffMs * 2, MAX_BACKOFF_MS);
          continue;
        }

        return;
      }

      throw new Error('Connection closed, cannot reconnect');
    })();

    try {
      await this.reconnectPromise;
    } finally {
      this.reconnectPromise = null;
    }
  }
}
